### ISSUE TITLE
lxc.rootfs.path containing symlinks not working

### ISSUE THREAD

**Issue by gjaekel** [lxc/lxc#4489]
# Issue description

TLDR: Starting a container with a rootfs-path containing a (neested?) symlink fail with recent _LXC_

I'm using LXC for decades since _LXC-0.8.4_. I'm running Gentoo-based containers on a bunch of Gentoo-based "diskless" bladeservers as hosts. The whole environment is using NFS for all the filesystem and all of the containers may be started on any of the hosts. The management is done by own scripts just using _LXC_ (not _LXD_).

This days I'm switching to _LXC-6.0.2_ from _LXC-3.1.2_ (yeah, finally! :) ). In general, I noticed not any serious problems for now - except for one issue: In contrast to before, `lxc-start` complain about *symlinks* inside the path to the containers rootfs (`lxc.rootfs.path`). In my starter script, this information is dynamically derived from a templated configuration data storage and (as almost every parameter) passed to `lxc-start` via commandline option (`-s lxc.rootfs.path=$ROOTPATH`). I'm very common using symlink to have a *centralized* mapping between abstract and concrete values

# Example

* To start e.g the container _gentest_, the passed value to `lxc.rootfs.path` is `/etc/lxc/gentemp/rootfs`
* This is a symlink to `/import/.diskless/roots/gentemp` where
  * `.diskless` is a symlink to `.main/diskless`, where
    * `.main` is a symlink to `/mnt/.netapp2_2`, where
      * `.netapp2_2` is a symlink to `netapp2\02\q`
* therefore, `/etc/lxc/gentemp/rootfs` expands to `/mnt/netapp2/02/q/diskless/roots/gentest`

Using _LXC-6.0.2_, the start fails with

```
[...]
lxc-start gentemp 20241022091813.391 ERROR    dir - ../lxc-6.0.2/src/lxc/storage/dir.c:dir_mount:163 - Not a directory - Failed to open "/etc/lxc/gentemp/rootfs"
lxc-start gentemp 20241022091813.391 ERROR    conf - ../lxc-6.0.2/src/lxc/conf.c:lxc_mount_rootfs:1236 - Failed to mount rootfs "/etc/lxc/gentemp/rootfs" onto "/run/lxc" with options "(null)"
lxc-start gentemp 20241022091813.391 ERROR    conf - ../lxc-6.0.2/src/lxc/conf.c:lxc_setup_rootfs_prepare_root:3496 - Failed to setup rootfs for
lxc-start gentemp 20241022091813.391 ERROR    conf - ../lxc-6.0.2/src/lxc/conf.c:lxc_setup:3871 - Failed to setup rootfs
lxc-start gentemp 20241022091813.391 ERROR    start - ../lxc-6.0.2/src/lxc/start.c:do_start:1273 - Failed to setup container "gentemp"
[...]
```

But if I workaround this by expanding the path (by using something similar to `-s lxc.rootfs.path=$(readlink -f $ROOTPATH)`, all is working fine.

# Required information

 * Distribution: Gentoo Linux 
 * Distribution version: Portage Tree at about 20241014
 * The output of
   * `lxc-start --version`
```
     6.0.2
```
   * `lxc-checkconfig`
```
root@genless ~ # lxc-checkconfig
LXC version 6.0.2

--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Network namespace: enabled
Namespace limits:
  cgroup: 63084
  ipc: 63084
  mnt: 63084
  net: 63084
  pid: 63084
  time: 63084
  user: 63084
  uts: 63084

--- Control groups ---
Cgroups: enabled
Cgroup namespace: enabled
Cgroup v1 mount points:
 - /sys/fs/cgroup/openrc
 - /sys/fs/cgroup/cpuset
 - /sys/fs/cgroup/cpu
 - /sys/fs/cgroup/cpuacct
 - /sys/fs/cgroup/blkio
 - /sys/fs/cgroup/memory
 - /sys/fs/cgroup/devices
 - /sys/fs/cgroup/freezer
 - /sys/fs/cgroup/net_cls
 - /sys/fs/cgroup/perf_event
 - /sys/fs/cgroup/net_prio
 - /sys/fs/cgroup/pids
 - /sys/fs/cgroup/misc
Cgroup v2 mount points:
 - /sys/fs/cgroup/unified
Cgroup v1 systemd controller: missing
Cgroup v1 clone_children flag: enabled
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled, not loaded
Macvlan: enabled, not loaded
Vlan: enabled, not loaded
Bridges: enabled, not loaded
Advanced netfilter: enabled, not loaded
CONFIG_IP_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled, not loaded
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled, not loaded
FUSE (for use with lxcfs): enabled, loaded

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: enabled
```
   * `uname -a`
```
Linux genless 6.6.56-gentoo #1 SMP PREEMPT_DYNAMIC Fri Oct 18 12:08:40 CEST 2024 x86_64 Intel(R) Core(TM) i5-8400 CPU @ 2.80GHz GenuineIntel GNU/Linux
```
   * `cat /proc/self/cgroup`
```
13:misc:/
12:pids:/
11:net_prio:/
10:perf_event:/
9:net_cls:/
8:freezer:/
7:devices:/
6:memory:/
5:blkio:/
4:cpuacct:/
3:cpu:/
2:cpuset:/
1:name=openrc:/sshd
0::/openrc.sshd 
```
   * `cat /proc/1/mounts`
```
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
udev /dev devtmpfs rw,nosuid,noexec,relatime,size=10240k,nr_inodes=2018708,mode=755 0 0
devpts /dev/pts devpts rw,relatime,gid=5,mode=620,ptmxmode=000 0 0
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
A.B.C.D:/02/q/diskless/roots/genless / nfs rw,relatime,vers=3,rsize=1048576,wsize=1048576,namlen=255,hard,nolock,proto=tcp,timeo=600,retrans=2,sec=sys,mountaddr=A.B.C.D,mountvers=3,mountproto=tcp,local_lock=all,addr=A.B.C.D 0 0
tmpfs /run tmpfs rw,nosuid,nodev,size=3231076k,nr_inodes=819200,mode=755 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
debugfs /sys/kernel/debug debugfs rw,nosuid,nodev,noexec,relatime 0 0
efivarfs /sys/firmware/efi/efivars efivarfs rw,nosuid,nodev,noexec,relatime 0 0
cgroup_root /sys/fs/cgroup tmpfs rw,nosuid,nodev,noexec,relatime,size=10240k,mode=755 0 0
openrc /sys/fs/cgroup/openrc cgroup rw,nosuid,nodev,noexec,relatime,release_agent=/lib/rc/sh/cgroup-release-agent.sh,name=openrc 0 0
none /sys/fs/cgroup/unified cgroup2 rw,nosuid,nodev,noexec,relatime 0 0
cpuset /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset,clone_children 0 0
cpu /sys/fs/cgroup/cpu cgroup rw,nosuid,nodev,noexec,relatime,cpu 0 0
cpuacct /sys/fs/cgroup/cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpuacct 0 0
blkio /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
memory /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory,clone_children 0 0
devices /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
freezer /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
net_cls /sys/fs/cgroup/net_cls cgroup rw,nosuid,nodev,noexec,relatime,net_cls 0 0
perf_event /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
net_prio /sys/fs/cgroup/net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_prio 0 0
pids /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
misc /sys/fs/cgroup/misc cgroup rw,nosuid,nodev,noexec,relatime,misc 0 0
mqueue /dev/mqueue mqueue rw,nosuid,nodev,noexec,relatime 0 0
shm /dev/shm tmpfs rw,nosuid,nodev,noexec,relatime 0 0
portage-tmp /var/tmp/portage tmpfs rw,relatime,size=8388608k,nr_inodes=1048576 0 0
/dev/sda8 /mnt/netapp2 ext4 rw,relatime 0 0
rpc_pipefs /var/lib/nfs/rpc_pipefs rpc_pipefs rw,relatime 0 0
netapp2:/ /mnt/netapp2 nfs rw,relatime,vers=3,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,mountaddr=A.B.C.D,mountvers=3,mountport=635,mountproto=udp,fsc,local_lock=none,addr=A.B.C.D 0 0
ess1:/ibm/gpfs1/export0 /mnt/ess1_0 nfs rw,relatime,vers=3,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,nconnect=4,timeo=10,retrans=18,sec=sys,mountaddr=A.B.C.D,mountvers=3,mountport=58106,mountproto=udp,fsc,local_lock=none,addr=A.B.C.D 0 0
netapp2:/02 /mnt/netapp2/02 nfs rw,relatime,vers=3,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,fsc,local_lock=none,addr=A.B.C.D 0 0
netapp4:/ /mnt/netapp4 nfs rw,relatime,vers=3,rsize=65536,wsize=65536,namlen=255,soft,proto=tcp,nconnect=4,timeo=600,retrans=2,sec=sys,mountaddr=10.34.63.196,mountvers=3,mountport=635,mountproto=udp,fsc,local_lock=none,addr=10.34.63.196 0 0
lxcfs /var/lib/lxcfs fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
netapp2:/02/q/diskless/roots/kickstart0/import/.inject /mnt/netapp2/02/q/diskless/roots/kickstart0/import/.inject nfs rw,relatime,vers=3,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,fsc,local_lock=none,addr=A.B.C.D 0 0
netapp2:/02/q/diskless/roots/_template.lxc-client.plain/import/.inject /mnt/netapp2/02/q/diskless/roots/_template.lxc-client.plain/import/.inject nfs rw,relatime,vers=3,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,fsc,local_lock=none,addr=A.B.C.D 0 0 
```

**Comment by mihalicyn** [comment]
Hi @gjaekel,

this behavior is intended. A while ago a lot of hardening was done on LXC side to prevent potential vulnerabilities, because symlinks can be extremely dangerous in some circumstances.

> But if I workaround this by expanding the path (by using something similar to `-s lxc.rootfs.path=$(readlink -f $ROOTPATH)`, all is working fine.

I believe that expanding a symlink in a controlled way is a good way to go here, unless this `$ROOTPATH` this doesn't come from untrusted source.

As this was opened without activity from our side for a long time (sorry about this), I'm closing this for now. Feel free to reopen or leave comment if you want to discuss this further.
